modules/sw360/sw360-workflow/src/main/java/org/eclipse/sw360/antenna/sw360/SW360MetaDataUpdater.java

```diff
@@ -108,15 +112,74 @@ public boolean isUploadSources() {
-        LOGGER.debug("Result of attachment upload operation: {}", result);
-        if (!result.isSuccess()) {
-            LOGGER.error("Failed to upload attachments: {}", result.failedUploads());
+        AttachmentUploadRequest<SW360Release> uploadRequest = builder.build();
+        if (!uploadRequest.getItems().isEmpty()) {
+            AttachmentUploadResult<SW360Release> result = releaseClientAdapter.uploadAttachments(uploadRequest);
+            LOGGER.debug("Result of attachment upload operation: {}", result);
+            if (!result.isSuccess()) {
+                LOGGER.error("Failed to upload attachments: {}", result.failedUploads());
+            }
+            return result.getTarget();
+        }
+        return sw360Release;
+    }
+
+    /**
+     * Calculates the SHA-1 hash for the given local attachment file. This is
+     * used to determine whether a modified attachment file needs to be
+     * uploaded.
+     *
+     * @param path the path to the local file
+     * @return the SHA-1 hash of this file
+     * @throws SW360ClientException if an error occurs
+     */
+    String calculateAttachmentHash(Path path) {
+        return SW360AttachmentUtils.calculateSha1Hash(path);
+    }
+
+    /**
+     * Checks whether an attachment upload is necessary for the given
+     * attachment file. This is the case if no attachment with this file name
+     * and type and identical content exists.
+     *
+     * @param release        the release affected
+     * @param path           the path to the local attachment file
+     * @param attachmentType the attachment type
+     * @return a flag whether this attachment file must be uploaded
+     */
+    private boolean needUpload(SW360Release release, Path path, SW360AttachmentType attachmentType) {
+        String fileName = String.valueOf(path.getFileName());
+        Optional<SW360SparseAttachment> optAttachment = release.getEmbedded().getAttachments().stream()
+                .filter(attachment -> attachmentMatches(path, attachmentType, fileName, attachment))
+                .findFirst();
+        return !optAttachment.isPresent();
+    }
+
+    /**
+     * Checks whether an attachment associated with a release matches the given
+     * properties. If there is an exact match, the upload of this attachment
+     * file can be skipped.
+     *
+     * @param path           the path to the local attachment file
+     * @param attachmentType the attachment type
+     * @param fileName       the attachment file name
+     * @param attachment     the attachment from the current release
+     * @return a flag whether the attachment is matched by these properties
+     */
+    private boolean attachmentMatches(Path path, SW360AttachmentType attachmentType, String fileName, SW360SparseAttachment attachment) {
+        try {
+            return Objects.equals(fileName, attachment.getFilename()) &&
```

Comment:
Since we have checked last week, the duplicate attachment function in SW360 currently still marks anattachment as duplicate when either the filename OR the sha1 is the same, I think this should be reflected as well. 

Or is it your aim to have a logical duplicate check here and rather catch the exception thrown in SW360 until they remove the filename check in their duplicate attachment statement. In that case checking for the filename would be superfluous, because an attachment can not be updated if for example only the sha1 matches. 

Either way I think this deserves a comment for future changes that it either doesn't reflect the duplicate checking in SW360 or it should reflect the duplicate checking of SW360

Reply:
The matches check now only checks for the SHA-1 value. A comment has been added that mentions the current limitations of the solution.

The final solution will probably depend on the functionality SW360 will offer in this area in future. I hope that it will be possible to delete all wrong or out-dated source attachments before uploading the correct one.

Reply:
Hm, I might not have said that right earlier. 
The duplicate attachment checks `(isFileNameDuplicate || isSha1Duplicate)`, so there still is a need to test the filename, since the OR condition will give back true if the filenames match

Reply:
I hope, this is covered by the comment which says that the result of the check method does not guarantee a successful upload. The compliance tool will probably have to do additional steps (like deleting old attachments) before the curated source attachment can be uploaded.

Reply:
okay, could you create an Issue for the TODO and link it in this PR? Then everything is properly documented

Reply:
Issue was created:
https://github.com/eclipse/antenna/issues/538